File: DecisionTrees.py
# -*- coding: utf-8 -*-
"""
Created on Wed Oct 12 10:20:37 2022

@author: samst
"""

#import normalizing
from sklearn import preprocessing
# Pandas for data frames
import pandas as pd
import numpy as np
# Decision Trees
from sklearn.tree import DecisionTreeClassifier
from sklearn import tree

# VIZ
import matplotlib.pyplot as plt
import graphviz 
from sklearn.metrics import confusion_matrix

############ Utility Functions ##############


### Function to output string version of row 
# Row is row that should be tablefied
# Length should be length of row not counting name
def makeRow(row, length, rowName):
    finalRow = "$" + rowName + "$"
    for i in range(0,length):
            finalRow += "&$" + str(row[i])+ "$"
    finalRow += "\\\\"
    return finalRow
# Function to give confusion matrix as latex table
# Data should be a 2 dimensional symmetric matrix
# Col names should be the names of the columns
# Row names should be the names of the rows in first column
def prettyLatexTables(data, colNames, rowNames, tableName):
    assert len(colNames) == len(rowNames), "colNames and rowNames should have same number of elements"
    size = len(colNames)
    numCols = ["c|" for col in colNames]
    numCols.append("c")
    table = ("\\begin{table}\\begin{tabular}{"+"".join(numCols)+"}" +
             makeRow(colNames, size, "True Value, Output of Model") +
             "\\hline" +
             "".join([makeRow(data[i], size, rowNames[i]) for i in range(0,size)])+
        "\\end{tabular}\n\centering\caption{"+tableName+"}\n"+
        "\end{table}")
    return table


##################################################


## Prep Work

path="C:\\Users\\samst\\Desktop\\MLProject\\MLProjectMod3\\"
# Latex output file
LATEXPATH = path + "\\Viz\\DTConfMats.txt"
LATEXFILE = open(LATEXPATH, "w")
LATEXFILE.close()

# load in dataset as pandas dataframe
trainData = pd.read_csv(path+"trainingData.csv")
testData  = pd.read_csv(path+"testingData.csv")

testLabels = testData["Label"]
trainLabels = trainData["Label"]

# Remove Labels
testData = testData.drop("Label",axis=1)
trainData = trainData.drop("Label",axis=1)
# First column is a primary key
testData = testData.drop("Unnamed: 0",axis=1)
trainData = trainData.drop("Unnamed: 0",axis=1)

# Cast non-numerical data types to numerical,
# This works for decision tree but not naive bayes
# Categorical WorkClass - Sex


encoder = preprocessing.OrdinalEncoder()

testData["WorkClass"] = encoder.fit_transform(np.array(testData["WorkClass"]).reshape(-1,1))
testData["Education"] = encoder.fit_transform(np.array(testData["Education"]).reshape(-1,1))
testData["MaritalStatus"] = encoder.fit_transform(np.array(testData["MaritalStatus"]).reshape(-1,1))
testData["Occupation"] = encoder.fit_transform(np.array(testData["Occupation"]).reshape(-1,1))
testData["RelationshipToFamily"] = encoder.fit_transform(np.array(testData["RelationshipToFamily"]).reshape(-1,1))
testData["Race"] = encoder.fit_transform(np.array(testData["Race"]).reshape(-1,1))
testData["Sex"] = encoder.fit_transform(np.array(testData["Sex"]).reshape(-1,1))


trainData["WorkClass"] = encoder.fit_transform(np.array(trainData["WorkClass"]).reshape(-1,1))
trainData["Education"] = encoder.fit_transform(np.array(trainData["Education"]).reshape(-1,1))
trainData["MaritalStatus"] = encoder.fit_transform(np.array(trainData["MaritalStatus"]).reshape(-1,1))
trainData["Occupation"] = encoder.fit_transform(np.array(trainData["Occupation"]).reshape(-1,1))
trainData["RelationshipToFamily"] = encoder.fit_transform(np.array(trainData["RelationshipToFamily"]).reshape(-1,1))
trainData["Race"] = encoder.fit_transform(np.array(trainData["Race"]).reshape(-1,1))
trainData["Sex"] = encoder.fit_transform(np.array(trainData["Sex"]).reshape(-1,1))
#############    Decision Tree   #######################

# Set up decision Tree
# Worked slightly better with gini impurity than entropy
decisionTree=DecisionTreeClassifier(criterion='gini', ##"entropy" or "gini"
                            splitter='best',  ## or "random" or "best"
                            max_depth=20, 
                            min_samples_split=2, 
                            min_samples_leaf=1, 
                            min_weight_fraction_leaf=0.0, 
                            max_features=None, 
                            random_state=None, 
                            max_leaf_nodes=None, 
                            min_impurity_decrease=0.001, # Set to reduce tree to manageable size
                            class_weight=None)

# Build model from training data
decisionTree.fit(trainData, trainLabels)
# Plot Tree
#tree.plot_tree(decisionTree)
featureNames=trainData.columns
dot_data = tree.export_graphviz(decisionTree, out_file=None,
                    ## The following creates TrainDF.columns for each
                    ## which are the feature names.
                      feature_names=featureNames,  
                      #class_names=MyDT.class_names,  
                      filled=True, rounded=True,  
                      special_characters=True)                                    
graph = graphviz.Source(dot_data) 
graph.render('Decision Tree') 


print("\nTrain Accuracy:")
prediction=decisionTree.predict(trainData)
confusionMatrix = confusion_matrix(trainLabels, prediction)
#### CONFUSION MATRIX TEXT EXPORT TO LATEX
LATEXFILE = open(LATEXPATH, "a")
LATEXFILE.write(prettyLatexTables(confusionMatrix, [">50k", "<=50k"],[">50k", "<=50k"], "Decision Tree Training Confusion Matrix"))
LATEXFILE.close()

accuracy = (confusionMatrix[0][0]+confusionMatrix[1][1])/(confusionMatrix[0][0]+confusionMatrix[0][1]+confusionMatrix[1][0]+confusionMatrix[1][1])
print(accuracy)
## Eval tree on test set


#print("Prediction\n")
prediction=decisionTree.predict(testData)
#print(prediction)
## Show the confusion matrix
confusionMatrix = confusion_matrix(testLabels, prediction)
print("\nThe confusion matrix is:")
print(confusionMatrix)
#### CONFUSION MATRIX TEXT EXPORT TO LATEX
LATEXFILE = open(LATEXPATH, "a")
LATEXFILE.write(prettyLatexTables(confusionMatrix, [">50k", "<=50k"],[">50k", "<=50k"], "Decision Tree Test Confusion Matrix"))
LATEXFILE.close()


print("\nTest Accuracy:")
accuracy = (confusionMatrix[0][0]+confusionMatrix[1][1])/(confusionMatrix[0][0]+confusionMatrix[0][1]+confusionMatrix[1][0]+confusionMatrix[1][1])
print(accuracy)

File: DecisionTrees.R
#install.packages('rpart')
#install.packages('rattle')

#Decision Trees
library(rpart)
library(rattle)  ## FOR Decision Tree Vis
setwd("C:\\Users\\samst\\Desktop\\MLProject\\MlPRojectMod3")
decisionTreeData <- read.csv("fullData.csv", header=TRUE, row.names=1)

testData  <- read.csv("testingData.csv", header=TRUE, row.names=1)
trainData <- read.csv("trainingData.csv", header=TRUE, row.names=1)

##### Unfortunately R always loads factors as char
decisionTreeData$Label                 <- as.factor(decisionTreeData$Label)
decisionTreeData$WorkClass             <- as.factor(decisionTreeData$WorkClass)
decisionTreeData$Education             <- as.factor(decisionTreeData$Education)
decisionTreeData$MaritalStatus         <- as.factor(decisionTreeData$MaritalStatus)
decisionTreeData$Occupation            <- as.factor(decisionTreeData$Occupation)
decisionTreeData$RelationshipToFamily  <- as.factor(decisionTreeData$RelationshipToFamily)
decisionTreeData$Race                  <- as.factor(decisionTreeData$Race)
decisionTreeData$Sex                   <- as.factor(decisionTreeData$Sex)


trainData$WorkClass            <- as.factor(trainData$WorkClass)
trainData$Education            <- as.factor(trainData$Education)
trainData$MaritalStatus        <- as.factor(trainData$MaritalStatus)
trainData$Occupation           <- as.factor(trainData$Occupation)
trainData$RelationshipToFamily <- as.factor(trainData$RelationshipToFamily)
trainData$Race                 <- as.factor(trainData$Race)
trainData$Sex                  <- as.factor(trainData$Sex)
trainData$Label                <- as.factor(trainData$Label)


testData$WorkClass             <- as.factor(testData$WorkClass)
testData$Education             <- as.factor(testData$Education)
testData$MaritalStatus         <- as.factor(testData$MaritalStatus)
testData$Occupation            <- as.factor(testData$Occupation)
testData$RelationshipToFamily  <- as.factor(testData$RelationshipToFamily)
testData$Race                  <- as.factor(testData$Race)
testData$Sex                   <- as.factor(testData$Sex)
testData$Label                 <- as.factor(testData$Label)

# Notice no char types only factor which is good
str(testData)
str(trainData)
##################


# Set the seed so the trees are reproducible
set.seed(20)
######## First Tree using mixed data ##################

# Remove Labels from test data
testLabels <- testData$Label
testData <- testData[1:9]

# Convert label to factor
trainData$Label <-as.factor(trainData$Label)

# Run rpart decision tree
# Uses entropy as opposed to Gini
tree <- rpart(trainData$Label ~ ., data = trainData, method="class", parms = list(split="information"))
summary(tree)

prediction= predict(tree, testData, type="class")
## Confusion Matrix
table(prediction,testLabels) ## one way to make a confusion matrix

# Compute Accuracies

(confMat <- as.data.frame(table(prediction,testLabels)))
"Test Accuracy"
(accuracy <- (confMat$Freq[1] + confMat$Freq[4])/sum(confMat$Freq))

"Train Accuracy"
prediction= predict(tree, trainData, type="class")
(confMat <- as.data.frame(table(prediction,trainData$Label)))
(accuracy <- (confMat$Freq[1] + confMat$Freq[4])/sum(confMat$Freq))

## VIS..................
fancyRpartPlot(tree)

#######################################################
######## Second Tree using discrete Data ##############
#######################################################
# We want to discretize:
# V1 = Age
# V13 = hours-per-week

# Age will be simple
# <=18  is a child
# 19-40 is adult
# 40-60 is middle aged
# 60+   is old
decisCopy <- decisionTreeData
decisionTreeData$'Age'[as.double(decisCopy$'Age') <   40] <- 'Adult'
decisionTreeData$'Age'[as.double(decisCopy$'Age') <=  18] <- 'Child'
decisionTreeData$'Age'[as.double(decisCopy$'Age') >=  40] <- 'Middle Aged'
decisionTreeData$'Age'[as.double(decisCopy$'Age') >=  60] <- 'Old'

# Hours worked will be based slightly on the legalities
# 35 - 50  is full time
# <35 is part time
# >50 is absurd
decisionTreeData$'Hours.Per.Week'[as.double(decisCopy$'Hours.Per.Week') <   50] <- 'Full Time'
decisionTreeData$'Hours.Per.Week'[as.double(decisCopy$'Hours.Per.Week') <   35] <- 'Part Time'
decisionTreeData$'Hours.Per.Week'[as.double(decisCopy$'Hours.Per.Week') >=  50] <- 'Absurd'

############# SAMPLING ###############
# We need to resample
# As we can see there are 3* as many people with <= 50k as there are >50k so we need to somehow take a balanced sample for training
# For testing it doesn't actually need to be balanced.
# To do this we undersample
lessThan50    <- which(decisionTreeData$'Label' == ' <=50K')
greaterThan50 <- which(decisionTreeData$'Label' == ' >50K')

# Since we had 7000 >50k points lets go for 10k sample size 5k of each 
n <- 10000
# Undersampling
lessThan50Sample    <- sample(lessThan50, floor(n/2))
greaterThan50Sample <- sample(greaterThan50, floor(n/2))
sample              <- c(lessThan50Sample, greaterThan50Sample)

# Disjoint test and training data
trainData <- decisionTreeData[sample, ]
testData  <- decisionTreeData[-sample, ]

# Remove Labels from test data
testLabels <- testData$'Label'
testData   <- subset (testData, select = -Label)

# Convert label to factor
trainData$'Label' <- as.factor(trainData$'Label')

################# Decision Tree##############
# Run rpart decision tree
# cp determines size of tree
tree <- rpart(trainData$Label ~ ., data = trainData, method="class", cp=0.012)
summary(tree)

prediction <- predict(tree, testData, type="class")
## Confusion Matrix
table(prediction,testLabels) ## one way to make a confusion matrix

# Compute Accuracy
confMat   <- as.data.frame(table(prediction,testLabels))
"Test Accuracy"
(accuracy <- (confMat$Freq[1] + confMat$Freq[4])/sum(confMat$Freq))

"Train Accuracy"
prediction <- predict(tree, trainData, type="class")
confMat    <- as.data.frame(table(prediction,trainData$'Label'))
(accuracy  <- (confMat$Freq[1] + confMat$Freq[4])/sum(confMat$Freq))


## VIS..................
fancyRpartPlot(tree)